# Step 1: Overview
### title:          proj03.py
### description:    CSC-125 Project 3: Profitability in Mobile App Development
### author:         jmalavasi
### date:           08012021
### version:        1.0

## Determining what makes apps profitable in iOS and Android
The goal of this project is to analyze two datasets from both major mobile app stores in order to explore what makes apps more profitable.

# Step 2 : Explore


Import the data and print some rows


In [19]:
import csv

# import raw data sets
google_data_raw = list(csv.reader(open('./googleplaystore.csv', encoding='utf8')))
google_data = google_data_raw[1:]
google_labels = google_data_raw[0]
apple_data_raw = list(csv.reader(open('./applestore.csv', encoding='utf8')))
apple_data = apple_data_raw[1:]
apple_labels = apple_data_raw[0]

print(google_labels,'\n',google_data[0:2],'\n\n\n')
print(apple_labels,'\n',apple_data[0:2])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 
 [['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', '7-Jan-18', '1.0.0', '4.0.3 and up'], ['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', '15-Jan-18', '2.0.0', '4.0.3 and up']] 



['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 
 [['284882215', 'Facebook', '389879808', 'USD', '0', '2974676', '212', '3.5', '3.5', '95', '4+', 'Social Networking', '37', '1', '29', '1'], ['389801252', 'Instagram', '113954816', 'USD', '0', '2161558', '1289', '4.5', '4', '10.23', '12+', 'Photo & Video', '37', '

Use Explore Data Function

In [144]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))



In [145]:
# Explore Android
print('Google:\n',google_labels,'\n')
explore_data(google_data, 0, 3, True)

# Explore Apple
print('\n\nApple:\n',apple_labels,'\n')
explore_data(apple_data, 0, 3, True)

Google:
 ['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', '7-Jan-18', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', '15-Jan-18', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', '1-Aug-18', '1.2.4', '4.0.3 and up']


Number of rows: 10840
Number of columns: 13


Apple:
 ['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 

['284882215'

There are 10841 rows in the Google dataset and 7197 rows in the Apple dataset. From the google data set the columns of interest appear to be Category, Rating, Reviews, and Price. For Apple : track_name, price, rating_count_tot, and user_rating.

# Step 3 : Remove Bad Data



Discussion indicates that row 10742 is missing a category and should be deleted

In [46]:
print(google_labels,'\n',google_data[10472])
del google_data[10472]
print(google_labels,'\n',google_data[10472])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 
 ['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', '11-Feb-18', '1.0.19', '4.0 and up', '']
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 
 ['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', '7-Aug-18', '6.06.14', '4.4 and up']


# Step 4 : Finding Duplicates

After close inspection of the google data, you will notice duplicate apps. One example of this is Google Ads.

In [146]:
for row in google_data:
    app = row[0]
    if app == 'Google Ads':
        print(row)

['Google Ads', 'BUSINESS', '4.3', '29313', '20M', '5,000,000+', 'Free', '0', 'Everyone', 'Business', '30-Jul-18', '1.12.0', '4.0.3 and up']
['Google Ads', 'BUSINESS', '4.3', '29313', '20M', '5,000,000+', 'Free', '0', 'Everyone', 'Business', '30-Jul-18', '1.12.0', '4.0.3 and up']
['Google Ads', 'BUSINESS', '4.3', '29331', '20M', '5,000,000+', 'Free', '0', 'Everyone', 'Business', '30-Jul-18', '1.12.0', '4.0.3 and up']


Show total number of duplicates

In [47]:
filter_duplicate_apps = []
filter_unique_apps = []

for row in google_data:
    app_name = row[0]
    if app_name in filter_unique_apps:
        filter_duplicate_apps.append(app_name)
    else:
        filter_unique_apps.append(app_name)
    
print('Number of duplicate apps:', len(filter_duplicate_apps))
print('\n')
print('Examples of duplicate apps:', filter_duplicate_apps[:15])

Number of duplicate apps: 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


We don't want to randomly remove duplicates in order to avoid outliers. Therefore we will keep the record with the most reviews.

# Step 5: Removing Duplicates

In [53]:
reviews_max = {}

# for each app in google data, get record with max number of reviews
for app in google_data:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

# validate        
print('Expected length:', len(google_data) - 1181)
print('Actual length:', len(reviews_max))

Expected length: 9659
Actual length: 9659


Use reviews_max dictionary as a reference to clean actual data set. 

Instead of removing the duplicates randomly, we will choose the record with the most number of reviews.

We will do this by looping through each app in the dataset and comparing the number of reviews in each record to the data we collected in the previous step. If the number of reivews in the record matches the one in the reviews_max collection, then the record will be added to android_clean. Each type an app is added to the dataset, it is recorded in already_added to prevent duplicates.


In [147]:
android_clean = []
already_added = []

for app in google_data:
    name = app[0]
    n_reviews = float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name) 

# validate           
explore_data(android_clean, 0, 5, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', '7-Jan-18', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', '1-Aug-18', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', '8-Jun-18', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', '20-Jun-18', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', '26-Mar-17', '1', '2.3 and up']


Number of rows: 9659
Number of columns: 13


## Repeat for iOS

In [122]:

reviews_max_apple = {}

# for each app in google data, get record with max number of reviews
for app in apple_data:
    name = app[1]
    n_reviews = int(app[5])
    
    if name in reviews_max_apple and reviews_max_apple[name] < n_reviews:
        reviews_max_apple[name] = n_reviews
    elif name not in reviews_max_apple:
        reviews_max_apple[name] = n_reviews

apple_clean = []
already_added_apple = []


for app in apple_data:
    name = app[1]
    n_reviews = int(app[5])

    if (reviews_max_apple[name] == n_reviews) and (name not in already_added):
        apple_clean.append(app)
        already_added_apple.append(name) 

# validate           
explore_data(apple_clean, 0, 5, True)

['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0', '1126879', '3594', '4', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


['324684580', 'Spotify Music', '132510720', 'USD', '0', '878563', '8253', '4.5', '4.5', '8.4.3', '12+', 'Music', '37', '5', '18', '1']


['343200656', 'Angry Birds', '175966208', 'USD', '0', '824451', '107', '4.5', '3', '7.4.0', '4+', 'Games', '38', '0', '10', '1']


['469369175', 'CSR Racing', '524803072', 'USD', '0', '677247', '2029', '4.5', '4.5', '4.0.1', '4+', 'Games', '37', '5', '10', '1']


['924373886', 'Crossy Road - Endless Arcade Hopper', '165471232', 'USD', '0', '669079', '1087', '4.5', '4.5', '1.5.4', '9+', 'Games', '38', '5', '13', '1']


Number of rows: 6867
Number of columns: 16


# Step 6: Remove non English Records

In [59]:
def is_english_app(string):
    for char in string:
        if ord(char) > 127:
            return False
    return True

print("Testing Names:\n")
test_names = ["Instagram","爱奇艺PPS -《欢乐颂2》电视剧热播", "Docs To Go™ Free Office Suite","Instachat 😜"]
for name in test_names:
    print(name,': ',is_english_app(name))

Testing Names:

Instagram :  True
爱奇艺PPS -《欢乐颂2》电视剧热播 :  False
Docs To Go™ Free Office Suite :  False
Instachat 😜 :  False


# Step 7: Update Filter to Allow for Emojis

In [62]:
def is_english_app(string):
    strikes = 0
    for char in string:
        if ord(char) > 127:
            strikes += 1
    if strikes > 3:
        return False
    else:
        return True

print("Testing Names:\n")
test_names = ["Instagram","爱奇艺PPS -《欢乐颂2》电视剧热播", "Docs To Go™ Free Office Suite","Instachat 😜"]
for name in test_names:
    print(name,': ',is_english_app(name))

Testing Names:

Instagram :  True
爱奇艺PPS -《欢乐颂2》电视剧热播 :  False
Docs To Go™ Free Office Suite :  True
Instachat 😜 :  True


Clean Dataset

In [123]:
android_english_apps = []
ios_english_apps = []

for row in android_clean:
    app_name = row[0]
    if is_english_app(app_name):
        android_english_apps.append(row)
        
for row in apple_clean:
    app_name =  row[0]
    if is_english_app(app_name):
        ios_english_apps.append(row)
        
explore_data(android_english_apps, 0, 5, True)
print('\n')
explore_data(ios_english_apps, 0, 5, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', '7-Jan-18', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', '1-Aug-18', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', '8-Jun-18', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', '20-Jun-18', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', '26-Mar-17', '1', '2.3 and up']


Number of rows: 9614
Number of columns: 13


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0', '1126879

# Step 8: Remove Paid Apps

If the price of the app is 0, do not include it in final dataset

In [124]:
google_data_final = []
apple_data_final = []

for row in android_english_apps:
    price = row[7]
    if price == '0':
        google_data_final.append(row)
        
for row in ios_english_apps:
    price = row[4]
    if price == '0':
        apple_data_final.append(row)

print(len(google_data_final))
print(len(apple_data_final))

8864
3793


This leaves us with the following:
- Google : 8864 apps
- Apple : 3793 apps

# Step 9: Find Most Common Genres

Due to the fact that it is much easier to get an app on the Google Play Store, we will take the following approach to development. First we will build a minimal viable product (MVP) on android and then monitor its response. If the app takes off after an initial introduction period of six months, further development including introduction to the iOS App Store will take place. We will specifically investigate the prime_genre column from the Apple dataset and the Genre and Category columns from Google.

# Step 10: Create Function for Frequency Tables

Create code to analyze

In [125]:
def freq_table(dataset, index):
    # declare variables
    output_percentages = {}
    categories = {}   
 
    # for each record, count categories
    for row in dataset:
        value = row[index]
        if value in categories:
            categories[value] += 1
        else:
            categories[value] = 1 

    # for each category, convert count to %
    for category in categories:
        output_percentages[category] = (categories[category] / len(dataset)) * 100
    return output_percentages

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Run analysis

In [126]:
print("Google Data:\n")
display_table(google_data_final, 1) # Category

Google Data:

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART

In [127]:
print("Google Data:\n")
display_table(google_data_final, -4) # Genre

Google Data:

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0

In [128]:
print("Apple Data:\n")
display_table(apple_data_final, -5) # Genre

Apple Data:

Games : 56.02425520696019
Entertainment : 8.19931452675982
Photo & Video : 4.165568151858687
Social Networking : 3.374637490113367
Education : 3.295544423938835
Shopping : 3.031900870023728
Utilities : 2.7682573161086212
Lifestyle : 2.372791985235961
Finance : 2.056419720537833
Health & Fitness : 1.9509622989717903
Sports : 1.8455048774057472
Music : 1.7664118112312155
Book : 1.7400474558397045
News : 1.3709464803585552
Travel : 1.2391247034010018
Productivity : 1.2391247034010018
Food & Drink : 0.9227524387028736
Weather : 0.7909306617453202
Navigation : 0.5009227524387029
Reference : 0.47455839704719216
Business : 0.4218296862641709
Catalogs : 0.23727919852359608
Medical : 0.21091484313208544


# Step 11: Analyze Frequency Tables

## Apple Data
After pulling the most popular app categories from the Apple dataset, we can see that the top three categories are games, entertainment and photos & video. However, games occupies a much larger portion of the market (almost 7x the next category)


## Google Data
When looking at the Google dataset, we can see that both app markets are quite different. The largest number of apps in a specific category on android is in "Tools". This shows a larger focus on productivity amongst Android users.

# Step 12: Calculate Average Number of Installs Per App Genre - Apple

In [143]:
categories_apple = freq_table(apple_data_final, -5)

for category in categories_apple:
    total,len_category = 0,0
    for app in apple_data_final:
        app_category = app[-5]
        if app_category == category:            
            n_ratings = float(app[5])
            total += n_ratings
            len_category += 1
    avg_n_ratings = total / len_category
    print(category, ':', avg_n_ratings)

Music : 56482.02985074627
Games : 11259.265882352942
Health & Fitness : 19176.64864864865
Weather : 45041.53333333333
Utilities : 12775.466666666667
Shopping : 17678.93913043478
Social Networking : 16065.515625
Navigation : 26664.736842105263
Lifestyle : 7718.066666666667
Sports : 18031.914285714287
Photo & Video : 12421.056962025317
Book : 8498.333333333334
Entertainment : 7398.6012861736335
Finance : 12532.115384615385
Travel : 12080.446808510638
Reference : 18680.666666666668
Education : 5982.2
Productivity : 18663.27659574468
News : 8581.711538461539
Food & Drink : 4524.8
Business : 4744.5
Catalogs : 1779.5555555555557
Medical : 459.75


## Recommend a Profile

When it comes to install data on iOS, it seems like there is still a focus on entertainment with a high number of installs in categories like Music, Sports and Games. 

# Step 13: Calculate Average Number of Installs Per App Genre - Google

In [132]:
display_table(google_data_final,5)

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


In [142]:
categories_google = freq_table(google_data_final, 1)

for category in categories_google:
    total, len_category = 0,0
    for app in google_data_final:
        category_app = app[1]
        if category_app == category:            
            n_installs = app[5]
            n_installs = (n_installs.replace(',', '')).replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_number_installs = total / len_category
    print(category, ':', avg_number_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

## Recommend a Profile

When it comes to install data on Google, the data shifts a bit from what we saw when looking into categorical data only. In fact there are similar high target install categories to iOS such as communication, social, photography and games. 

# CONCLUSION- APP PROFILE RECOMENDATION

In conclusion, I would recommend an app profile that combines some of the top categories across both platforms. For example, a social media app that hosted its' own game library. As more and more people steer away from some of the larger players in social media in order to improve their personal privacy, niche social media apps will take off.